### PR TITLE
XML Export

### PR DESCRIPTION
**Pull request by gofal**
Changes proposed in this pull request:
- DicomDatasets can be exported as XML
-
-

### CODE REVIEW

**Comment by anders9ustafsson** [comment]
Many thanks for this PR, Reinhard, it is of course an extremely welcome addition to *fo-dicom*. A few questions and remarks:

1. You seem to be following the XML format outlined in DICOM standard PS 3.19, Section A. Any deviations from that specification in your implementation?
2. Before I accept the PR, I would like you to add API documentation, at the very least on the public methods and types.
3. I also want you to look at the failing unit tests. The problem seems to be related to line endings, so it should hopefully be quite easy to workaround.

**Comment by gofal** [comment]
Hi,
I am not aware of any devariation. Before implementing this I was searching in internet for examples of dicom-xml-files and I orientated on these. I was not aware that is is defined in dicom standard. I will check if the implementation is conform to the standard.
Where shall I document the API?
I will take a look at the unit test. In my visual studio all tests except of 2 pass (they are about converting decimals to string with dot or comma)

**Comment by gofal** [comment]
The first Implementation exported the PN-Tags just as normal Values, but they should be exported as PersonName. This is changed now. The XML output should now be as described in dicom standard.
I also added documentation to class and public methods.

File: DICOM/Log/DicomXML.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
+﻿using Dicom;
+using Dicom.IO.Buffer;
+using System;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DICOM.Shared.Log
```
Comment (anders9ustafsson) [inline DICOM/Log/DicomXML.cs]: Please change namespace for the `DicomXML` class to `Dicom.Log`.

File: DICOM/Log/DicomXML.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    public static class DicomXML
+    {
+       
+        /// <summary>
+        /// Converts a <see cref="DicomDataset"/> to a XML-String
+        /// </summary>
+        /// <param name="dataset">The DicomDataset that is converted to XML-String</param>
+        public static string ConvertDicomToXML(DicomDataset dataset)
+        {
+            string _xmlString = DicomToXml(dataset);
```
Comment (anders9ustafsson) [inline DICOM/Log/DicomXML.cs]: Please do not use underline prefix for local variables.

File: DICOM/Log/DicomXML.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            return _xmlString;
+        }
+
+        /// <summary>
+        /// Converts a <see cref="DicomDataset"/> to a XML-String asynchronous 
+        /// </summary>
+        /// <param name="dataset"></param>
+        /// <returns></returns>
+        public static Task<string> ConvertDicomToXMLAsync(DicomDataset dataset)
```
Comment (anders9ustafsson) [inline DICOM/Log/DicomXML.cs]: Is this method really necessary? It is not asynchronous in practice, and there is no interfaces that enforces it to be included.

File: DICOM/Log/DicomXML.cs (hunk 1 above)
Comment (gofal) [inline DICOM/Log/DicomXML.cs]: done

File: DICOM/Log/DicomXML.cs (hunk 2 above)
Comment (gofal) [inline DICOM/Log/DicomXML.cs]: removed the local variables. There are other classes with underline prefixes, I removed it also in one of these

File: DICOM/Log/DicomXML.cs (hunk 3 above)
Comment (gofal) [inline DICOM/Log/DicomXML.cs]: No is not necessary. I thought that the async-pattern should be enforced so I added this method, but I removed it now.

**Comment by gofal** [comment]
Hm, don't understand why this one test failed. It says it refuses connection on Dicom.Network. But there was no change in Dicom.Network.

File: DICOM/Log/DicomDatasetLogger.cs [hunk 4]
```diff
@@ -17,17 +17,17 @@ namespace Dicom.Log
     /// </summary>
     public class DicomDatasetLogger : IDicomDatasetWalker
     {
-        private Logger _log;
```
Comment (anders9ustafsson) [inline DICOM/Log/DicomDatasetLogger.cs]: My comment on underscore prefix only concerned local variables, not fields. There currently are inconsistencies in the different files, I am fully aware of that, but for private fields underscore prefix is OK.

**Comment by anders9ustafsson** [comment]
I have left another line comment regarding underscore prefix. But you don't need to do anymore with this, I intend to accept the PR very soon. You should not be too worried about the failed unit test either, occasionally this might still happen.

I will make some final adjustments to the PR. All in all I am very grateful for you providing this valuable addition to this library.

Thanks!

**Comment by anders9ustafsson** [comment]
Reinhard, I have now pulled this PR.

I made the following adjustments:

* Reverted your modification of the `DicomDatasetLogger`.
* Moved `DicomXML` to the *Dicom.Serialization* namespace.
* Moved the extension methods into the `DicomXML` class, since the class itself is static.

Once again, many thanks for your invaluable work on this issue. It's really great to be able to claim XML export support in the 3.0 release.

**Comment by Zaid-Safadi** [comment]
Hi @gofal, this is nice work and addition to fo-dicom. 

If you also like to support reading an XML string back to a DicomDataset please feel free to use the code here [XmlDicomConverter.cs](https://github.com/Zaid-Safadi/DICOMcloud/blob/development/DICOMcloud.Dicom/XmlDicomConverter.cs) as you like. 

Given that your write/convert to XML complies with the DICOM standard the read method I have will work with your generated string with no issues.

**Pull request merged**